### PR TITLE
load balancers: implement host retry predicates for consistent hash LBs

### PR DESCRIPTION
**Pull request by zuercher**
For Maglev, the computed hash is manipulated to choose another host.
For a given hash and fixed host set, the Nth retry maps to a stable
host, though no necessarly a different host.

For ring-hash, the behavior is as if the given host were removed from
the ring and the request falls through to the subsequent host.

Risk Level: medium (existing configs may be affected)
Testing: added unit tests
Docs Changes: n/a
Release Notes: updated

Signed-off-by: Stephan Zuercher <zuercher@gmail.com>

### CODE REVIEW

**Comment by zuercher** [comment]
I marked this one medium because in theory an existing configuration that enables host selection predicates for a ring hash or maglev cluster would suddenly change behavior with this code. I think it's unlikely folks would enable those features when they weren't working, however.

File: test/common/upstream/maglev_lb_test.cc [hunk 1]
```diff
@@ -119,6 +132,56 @@ TEST_F(MaglevLoadBalancerTest, BasicWithHostName) {
...
+
+  // maglev: i=0 host=127.0.0.1:92
+  // maglev: i=1 host=127.0.0.1:94
+  // maglev: i=2 host=127.0.0.1:90
+  // maglev: i=3 host=127.0.0.1:91
+  // maglev: i=4 host=127.0.0.1:95
+  // maglev: i=5 host=127.0.0.1:90
+  // maglev: i=6 host=127.0.0.1:93
+  LoadBalancerPtr lb = lb_->factory()->create();
+  const std::vector<HostConstSharedPtr> expected_hosts{
```
Comment (snowp) [inline test/common/upstream/maglev_lb_test.cc]: should this test just reference `host_set.hosts_`? not sure if this `expected_hosts` vec adds much

File: source/common/upstream/ring_hash_lb.cc [hunk 2]
```diff
@@ -79,9 +81,17 @@ HostConstSharedPtr RingHashLoadBalancer::Ring::chooseHost(uint64_t h) const {
     }
 
     if (lowp > highp) {
-      return ring_[0].host_;
+      midp = 0;
+      break;
     }
   }
+
+  // If a retry host predicate is being applied, behave as if this host was not in the ring.
```
[the remark this answers is not in the document]
Comment (zuercher) [reply source/common/upstream/ring_hash_lb.cc]: Added a note.

**Comment by mattklein123** [comment]
`Code coverage 96.56954322484786 is lower than limit of 97.0`. Is that real?

/wait

**Pull request merged**